kong/dns/client.lua

```diff
@@ -0,0 +1,691 @@
+    broadcast = function(channel, data)
+      if not kong or not kong.worker_events then
+        return
+      end
+
+      local ok, err = kong.worker_events.post(ipc_source, channel, data)
+      if not ok then
+        log(ERR, PREFIX, "failed to post event '", ipc_source, "', '", channel, "': ", err)
+      end
+    end,
+  }
+
+  local cache, err = mlcache.new("dns_cache", "kong_dns_cache", {
+    ipc = ipc,
+    neg_ttl = opts.error_ttl or DEFAULT_ERROR_TTL,
+    -- 10000 is a reliable and tested value from the original library.
+    lru_size = opts.cache_size or 10000,
+    shm_locks = ngx.shared.kong_locks and "kong_locks",
+    resty_lock_opts = resty_lock_opts,
+  })
+
+  if not cache then
+    return nil, "could not create mlcache: " .. err
+  end
+
+  if opts.cache_purge then
+    cache:purge(true)
+  end
+
+  -- parse hosts
+  local hosts, hosts_cache = init_hosts(cache, opts.hosts)
+
+  return setmetatable({
+    cache = cache,
+    stats = {},
+    hosts = hosts,
+    r_opts = r_opts,
+    resolv = opts._resolv or resolv,
+    valid_ttl = opts.valid_ttl,
+    error_ttl = opts.error_ttl or DEFAULT_ERROR_TTL,
+    stale_ttl = opts.stale_ttl or DEFAULT_STALE_TTL,
+    enable_srv = enable_srv,
+    enable_ipv4 = enable_ipv4,
+    enable_ipv6 = enable_ipv6,
+    hosts_cache = hosts_cache,
+
+    -- TODO: Make the table readonly. But if `string.buffer.encode/decode` and
+    -- `pl.tablex.readonly` are called on it, it will become empty table.
+    --
+    -- quickly accessible constant empty answers
+    EMPTY_ANSWERS = {
+      errcode = EMPTY_RECORD_ERROR_CODE,
+      errstr = EMPTY_RECORD_ERROR_MESSAGE,
+      ttl = opts.error_ttl or DEFAULT_ERROR_TTL,
+    },
+  }, _MT)
+end
+
+
+local function process_answers(self, qname, qtype, answers)
+  local errcode = answers.errcode
+  if errcode then
+    answers.ttl = self.error_ttl
+    return answers
+  end
+
+  local processed_answers = {}
+
+  -- 0xffffffff for maximum TTL value
+  local ttl = math_min(self.valid_ttl or 0xffffffff, 0xffffffff)
+
+  for _, answer in ipairs(answers) do
+    answer.name = string_lower(answer.name)
+
+    if self.valid_ttl then
+      answer.ttl = self.valid_ttl
+    else
+      ttl = math_min(ttl, answer.ttl)
+    end
+
+    local answer_type = answer.type
+
+    if answer_type == qtype then
+      -- compatible with balancer, see https://github.com/Kong/kong/pull/3088
+      if answer_type == TYPE_AAAA then
+        answer.address = ipv6_bracket(answer.address)
+
+      elseif answer_type == TYPE_SRV then
+        answer.target = ipv6_bracket(answer.target)
+      end
+
+      table_insert(processed_answers, answer)
+    end
+  end
+
+  if table_isempty(processed_answers) then
+    log(DEBUG, PREFIX, "processed ans:empty")
+    return self.EMPTY_ANSWERS
+  end
+
+  log(DEBUG, PREFIX, "processed ans:", #processed_answers)
+
+  processed_answers.expire = now() + ttl
+  processed_answers.ttl = ttl
```

Comment:
Let's say the `self.valid_ttl` is `10`, and the minimal `answer.ttl` is `5`. So the `processed_answers.tttl` will be 5? Does that mean the `self.valid_ttl` didn't override the TTL?

Reply:
` self.valid_ttl` will override TTL

See Line 314:

```
314     if self.valid_ttl then
315       answer.ttl = self.valid_ttl
316     else
317       ttl = math_min(ttl, answer.ttl)
318     end
```

Reply:
Because we combine (1)ttl caculating logic and (2)logic of adding `[]` to ip  into one `for` loop. So it looks a bit convoluted

Reply:
What's `processed_answers.ttl` used for?